core/embed/rust/src/ui/model_tr/component/loader.rs

```diff
@@ -256,6 +257,92 @@ impl LoaderStyleSheet {
     }
 }
 
+pub struct ProgressLoader<T>
+where
+    T: StringType,
+{
+    loader: Child<Progress<T>>,
+    duration_ms: u32,
+    loader_frames: u32,
+    timer_token: Option<TimerToken>,
+}
+
+impl<T> ProgressLoader<T>
+where
+    T: StringType + Clone,
+{
+    const LOADER_FRAMES_DEFAULT: u32 = 20;
+
+    pub fn new(loader_description: T, duration_ms: u32) -> Self {
+        Self {
+            loader: Child::new(
+                Progress::new(false, loader_description).with_icon(theme::ICON_LOCK_SMALL),
+            ),
+            duration_ms,
+            loader_frames: Self::LOADER_FRAMES_DEFAULT,
+            timer_token: None,
+        }
+    }
+
+    pub fn with_loader_frames(mut self, loader_frames: u32) -> Self {
+        self.loader_frames = loader_frames;
+        self
+    }
+
+    fn loader_refresh_duration(&self) -> Duration {
+        let loader_refresh_ms: u32 = self.duration_ms / self.loader_frames;
+        Duration::from_millis(loader_refresh_ms)
+    }
+
+    fn loader_refresh_step(&self) -> u16 {
+        LOADER_MAX / self.loader_frames as u16
+    }
+
+    pub fn start(&mut self, ctx: &mut EventCtx) {
+        self.timer_token = Some(ctx.request_timer(self.loader_refresh_duration()));
```

Comment:
Is it easy to do so? I was afraid that I would have to remember the starting time and advancing it on every frame by determining the current time (I never used `Instant::now()`, so that is my "fear")

... but I will try it, it could simplify things

Reply:
Done in [78ebc24](https://github.com/trezor/trezor-firmware/pull/3440/commits/78ebc24242251b295dab99a385547849858fea52)